src/stateful-event-subscriber.ts

```diff
@@ -20,13 +32,112 @@ export abstract class StatefulEventSubscriber<State>
   protected invalid: boolean = false;
 
   isTracking: () => boolean = () => false;
+  public addressesSubscribed: string[] = [];
+
+  private mapKey: string;
+  private cacheName: string;
 
-  constructor(public readonly name: string, protected logger: Logger) {}
+  public name: string;
+
+  constructor(
+    public readonly parentName: string,
+    _name: string,
+    protected dexHelper: IDexHelper,
+    protected logger: Logger,
+    private masterPoolNeeded: boolean = false,
+  ) {
+    this.name = _name.toLowerCase();
+    this.mapKey =
+      `${CACHE_PREFIX}_${this.dexHelper.config.data.network}_${this.parentName}`.toLowerCase();
+    this.cacheName = `${this.mapKey}_${this.name}`.toLowerCase();
+  }
 
   getStateBlockNumber(): Readonly<number> {
     return this.stateBlockNumber;
   }
 
+  //Function which set the initiale state and bounded it to blockNumber
+  //There is multiple possible case:
+  // 1. You provide a state in options object the function will initialize with the provided state
+  //  with blockNumber and subscribe to logs.
+  // 2. if you are a master instance of dex-lib and no state is provided in options object
+  //  then the function generate a new state with blockNumber as height and set the state with
+  //  the result.
+  // 3. if you are a slave instance of dex-lib
+  //  either:
+  //    - If a state is found in the cache and the state is not null we set our state with the
+  //      cache state and cache blockNumber. Subscribe to logs with the cache blockNumber
+  //  or:
+  //    - If no valid state found in cache, we generate a new state with blockNumber
+  //      and se state with blockNumber. Subscribe to logs with blockNumber. The function
+  //      will also publish a message to cache to tell one master version of dex-lib that this slave
+  //      instance subscribed to a pool from dex this.parentName and name this.name.
+  async initialize(
+    blockNumber: number,
+    options?: InitializeStateOptions<State>,
+  ) {
+    if (options && options.state) {
+      this.setState(options.state, blockNumber);
+    } else {
+      if (this.dexHelper.config.isSlave && this.masterPoolNeeded) {
```

Comment:
I don't fully understand property `this.masterPoolNeeded`. What is it used for?

Reply:
Create a pool in the master version of dex-lib.